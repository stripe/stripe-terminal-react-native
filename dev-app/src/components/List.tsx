import React from 'react';
import { ActivityIndicator, StyleSheet, Text, View } from 'react-native';
import { colors } from '../colors';

type Props = {
  title?: string;
  description?: string;
  loading?: boolean;
  children: React.ReactElement | React.ReactElement[];
  topSpacing?: boolean;
  bolded?: boolean;
  visible?: boolean;
};

export default function List({
  title,
  children,
  description,
  loading,
  topSpacing = true,
  bolded = true,
  visible = true,
}: Props) {
  if (!visible) {
    return <></>;
  }
  return (
    <View style={[styles.container, topSpacing && styles.topSpacing]}>
      <View style={styles.titleContainer}>
        {title && (
          <Text style={[styles.title, bolded && styles.bolded]}>{title}</Text>
        )}
        {description ? (
          <Text style={styles.description}>{description}</Text>
        ) : (
          loading && (
            <View style={styles.description}>
              <ActivityIndicator />
            </View>
          )
        )}
      </View>

      {children}
    </View>
  );
}

const styles = StyleSheet.create({
  title: {
    color: colors.dark_gray,
    paddingLeft: 16,
    marginVertical: 12,
  },
  bolded: {
    fontWeight: '600',
  },
  container: {
    marginBottom: 0,
  },
  description: {
    color: colors.dark_gray,
    paddingRight: 16,
  },
  titleContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  topSpacing: {
    marginTop: 22,
  },
});
