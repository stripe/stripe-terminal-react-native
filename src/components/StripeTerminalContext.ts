import { createContext } from 'react';
import type { Reader, InitParams, InitializeResultType } from '../types';

type ContextType = {
  loading: boolean;
  isInitialized: boolean;
  connectedReader?: Reader.Type | null;
  discoveredReaders: Reader.Type[];
  setLoading(value: boolean): void;
  setIsInitialized(value: boolean): void;
  setConnectedReader(value: Reader.Type | null): void;
  setDiscoveredReaders(value: Reader.Type[]): void;
  initialize?(params: InitParams): Promise<InitializeResultType>;
  log(code: string, message?: any): void;
};

export const StripeTerminalContext = createContext<ContextType>({
  discoveredReaders: [],
  loading: false,
  isInitialized: false,
  log: () => {},
  setLoading: () => {},
  setIsInitialized: () => {},
  setConnectedReader: () => {},
  initialize: undefined,
  setDiscoveredReaders: () => {},
});
