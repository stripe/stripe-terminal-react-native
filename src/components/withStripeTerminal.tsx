import React from 'react';
import { useStripeTerminal } from '../index';

export type WithStripeTerminalProps = ReturnType<typeof useStripeTerminal>;

/**
 *  withStripeTerminal HoC Component
 *
 * @example
 * ```ts
 *  function YourScreenComponent(props: WithStripeTerminalProps) { }
 *
 *  export default withStripeTerminal(YourScreenComponent);
 * ```
 * @param __namedParameters WithStripeTerminalProps
 * @returns React.JSX.Element
 * @category ReactComponents
 */
export function withStripeTerminal<Props>(
  WrappedComponent: React.ComponentType<Props>
) {
  const displayName =
    WrappedComponent.displayName || WrappedComponent.name || 'Component';

  const ComponentWithStripeTerminal = (props: Props) => {
    const terminalProps = useStripeTerminal();

    return <WrappedComponent {...terminalProps} {...props} />;
  };

  ComponentWithStripeTerminal.displayName = `withStripeTerminal(${displayName})`;

  return ComponentWithStripeTerminal;
}
