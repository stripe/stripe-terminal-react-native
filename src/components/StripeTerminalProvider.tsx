import React, { useCallback, useRef, useState, useMemo } from 'react';
import {
  Reader,
  LogLevel,
  StripeError,
  PaymentStatus,
  UserCallbacks,
  CommonError,
} from '../types';
import { StripeTerminalContext } from './StripeTerminalContext';
import { initialize, setConnectionToken } from '../functions';
import { useListener } from '../hooks/useListener';
import { NativeModules } from 'react-native';
import EventEmitter from 'react-native/Libraries/vendor/emitter/EventEmitter';

const {
  FETCH_TOKEN_PROVIDER,
  CHANGE_CONNECTION_STATUS,
  CHANGE_PAYMENT_STATUS,
  FINISH_DISCOVERING_READERS,
  FINISH_INSTALLING_UPDATE,
  REQUEST_READER_DISPLAY_MESSAGE,
  REQUEST_READER_INPUT,
  REPORT_AVAILABLE_UPDATE,
  REPORT_UNEXPECTED_READER_DISCONNECT,
  REPORT_UPDATE_PROGRESS,
  START_INSTALLING_UPDATE,
  UPDATE_DISCOVERED_READERS,
} = NativeModules.StripeTerminalReactNative.getConstants();

const emitter = new EventEmitter();

const TOKEN_PROVIDER_ERROR_MESSAGE =
  "Couldn't fetch connection token. Please check your tokenProvider method";

/**
 * @ignore
 */
type EventResult<T> = {
  result: T;
};

/**
 *  StripeTerminalProvider Component Props
 */
export interface Props {
  children: React.ReactElement | React.ReactElement[];
  /*
   You must set a token provider before calling any Stripe terminal method.
   This should fetch a connection token from your server and return it as a string.
    * @example
    * ```ts
    * const fetchTokenProvider = async () => {
    *   const response = await fetch('http://api_url/connection_token');
    *   const { secret } = await response.json();
    *   return secret;
    * };
    * ```
   */
  tokenProvider: () => Promise<string>;
  logLevel?: LogLevel;
}

/**
 *  StripeTerminalProvider Component
 *
 * @example
 * ```ts
 *  <StripeTerminalProvider tokenProvider={tokenProvider}>
 *    <App />
 *  </StripeTerminalProvider>
 * ```
 * @param __namedParameters Props
 * @returns JSX.Element
 * @category ReactComponents
 */
export function StripeTerminalProvider({
  children,
  tokenProvider,
  logLevel,
}: Props) {
  const [isInitialized, setIsInitialized] = useState(false);
  const [loading, setLoading] = useState(true);
  const [connectedReader, setConnectedReader] = useState<Reader.Type | null>();
  const [discoveredReaders, setDiscoveredReaders] = useState<Reader.Type[]>([]);
  const userCallbacks = useRef<UserCallbacks>();

  const log = useCallback(
    (code: string, message?: any) => {
      if (logLevel === 'verbose') {
        console.log(`[Stripe terminal]: ${code}`, message);
      }
    },
    [logLevel]
  );

  const tokenProviderHandler = async () => {
    try {
      const connectionToken = await tokenProvider();

      setConnectionToken(connectionToken);
    } catch (error) {
      setConnectionToken(undefined, TOKEN_PROVIDER_ERROR_MESSAGE);

      console.error(error);
      console.error(TOKEN_PROVIDER_ERROR_MESSAGE);
    }
  };

  const didUpdateDiscoveredReaders = useCallback(
    ({ readers }: { readers: Reader.Type[] }) => {
      log('Discovered readers', readers);

      setDiscoveredReaders(readers);
      userCallbacks.current?.onUpdateDiscoveredReaders?.(readers);
      emitter.emit(UPDATE_DISCOVERED_READERS);
    },
    [setDiscoveredReaders, log]
  );

  const didFinishDiscoveringReaders = useCallback(
    ({ result }: EventResult<{ error?: StripeError }>) => {
      if (result.error) {
        const { error } = result;
        log(
          'Discovering readers has been finished with the following error:',
          `code: ${error.code}, message: ${error.message}`
        );
      }
      userCallbacks.current?.onFinishDiscoveringReaders?.(result.error);
      emitter.emit(FINISH_DISCOVERING_READERS);
    },
    [log]
  );

  const didReportUnexpectedReaderDisconnect = useCallback(
    ({ error }: { error?: StripeError }) => {
      if (error) {
        log(`code: ${error.code}, message: ${error.message}`);
      }
      setConnectedReader(null);
      setDiscoveredReaders([]);
      userCallbacks.current?.onDidReportUnexpectedReaderDisconnect?.(error);
      emitter.emit(REPORT_UNEXPECTED_READER_DISCONNECT);
    },
    [log, setConnectedReader, setDiscoveredReaders]
  );

  const didReportAvailableUpdate = useCallback(
    ({ result }: EventResult<Reader.SoftwareUpdate>) => {
      log('didReportAvailableUpdate', result);
      userCallbacks.current?.onDidReportAvailableUpdate?.(result);
      emitter.emit(REPORT_AVAILABLE_UPDATE);
    },
    [log]
  );

  const didStartInstallingUpdate = useCallback(
    ({ result }: EventResult<Reader.SoftwareUpdate>) => {
      log('didStartInstallingUpdate', result);
      userCallbacks.current?.onDidStartInstallingUpdate?.(result);
      emitter.emit(START_INSTALLING_UPDATE);
    },
    [log]
  );

  const didReportReaderSoftwareUpdateProgress = useCallback(
    ({ result }: EventResult<{ progress: string }>) => {
      log('didReportReaderSoftwareUpdateProgress', result);
      userCallbacks.current?.onDidReportReaderSoftwareUpdateProgress?.(
        result.progress
      );
      emitter.emit(REPORT_UPDATE_PROGRESS);
    },
    [log]
  );

  const didFinishInstallingUpdate = useCallback(
    ({
      result,
    }: EventResult<Reader.SoftwareUpdate | { error: StripeError }>) => {
      if ((result as { error: StripeError }).error) {
        const { error } = result as { error: StripeError };
        log(
          'Install update failed with the following error:',
          `code: ${error.code}, message: ${error.message}`
        );
        userCallbacks.current?.onDidFinishInstallingUpdate?.({
          update: undefined,
          error: error,
        });
      } else {
        log('didFinishInstallingUpdate', result);
        userCallbacks.current?.onDidFinishInstallingUpdate?.({
          update: result as Reader.SoftwareUpdate,
          error: undefined,
        });
      }
      emitter.emit(FINISH_INSTALLING_UPDATE);
    },
    [log]
  );

  const didRequestReaderInput = useCallback(
    ({ result }: EventResult<Reader.InputOptions[]>) => {
      log('didRequestReaderInput', result);
      userCallbacks.current?.onDidRequestReaderInput?.(result);
      emitter.emit(REQUEST_READER_INPUT);
    },
    [log]
  );

  const didRequestReaderDisplayMessage = useCallback(
    ({ result }: EventResult<Reader.DisplayMessage>) => {
      log('didRequestReaderDisplayMessage', result);
      userCallbacks.current?.onDidRequestReaderDisplayMessage?.(result);
      emitter.emit(REQUEST_READER_DISPLAY_MESSAGE);
    },
    [log]
  );

  const didChangePaymentStatus = useCallback(
    ({ result }: EventResult<PaymentStatus>) => {
      log('didChangePaymentStatus', result);
      userCallbacks.current?.onDidChangePaymentStatus?.(result);
      emitter.emit(CHANGE_PAYMENT_STATUS);
    },
    [log]
  );

  const didChangeConnectionStatus = useCallback(
    ({ result }: EventResult<Reader.ConnectionStatus>) => {
      log('didChangeConnectionStatus', result);
      userCallbacks.current?.onDidChangeConnectionStatus?.(result);
      emitter.emit(CHANGE_CONNECTION_STATUS);
    },
    [log]
  );

  useListener(FETCH_TOKEN_PROVIDER, tokenProviderHandler);
  useListener(REPORT_AVAILABLE_UPDATE, didReportAvailableUpdate);
  useListener(START_INSTALLING_UPDATE, didStartInstallingUpdate);
  useListener(REPORT_UPDATE_PROGRESS, didReportReaderSoftwareUpdateProgress);
  useListener(FINISH_INSTALLING_UPDATE, didFinishInstallingUpdate);

  useListener(UPDATE_DISCOVERED_READERS, didUpdateDiscoveredReaders);
  useListener(FINISH_DISCOVERING_READERS, didFinishDiscoveringReaders);
  useListener(
    REPORT_UNEXPECTED_READER_DISCONNECT,
    didReportUnexpectedReaderDisconnect
  );
  useListener(REQUEST_READER_INPUT, didRequestReaderInput);
  useListener(REQUEST_READER_DISPLAY_MESSAGE, didRequestReaderDisplayMessage);
  useListener(CHANGE_PAYMENT_STATUS, didChangePaymentStatus);
  useListener(CHANGE_CONNECTION_STATUS, didChangeConnectionStatus);

  const setUserCallbacks = useCallback((callbacks: UserCallbacks) => {
    userCallbacks.current = callbacks;
  }, []);

  const _initialize = useCallback(async () => {
    setLoading(true);

    // test tokenProvider method since native SDK's doesn't fetch it on init
    try {
      await tokenProvider();
    } catch (error) {
      console.error(TOKEN_PROVIDER_ERROR_MESSAGE);
      console.error(error);

      return {
        error: {
          code: CommonError.Failed,
          message: TOKEN_PROVIDER_ERROR_MESSAGE,
        },
      };
    }

    const response = await initialize({ logLevel });

    if (response.error) {
      log(response.error.code, response.error.message);
    } else if (response.reader) {
      log('Connected to the reader: ', response.reader);
      setConnectedReader(response.reader);
    }

    if (!response.error) {
      setIsInitialized(true);
    }

    setLoading(false);

    return response;
  }, [logLevel, tokenProvider, log]);

  const value = useMemo(
    () => ({
      loading,
      isInitialized,
      connectedReader,
      discoveredReaders,
      setIsInitialized,
      setLoading,
      setConnectedReader,
      setDiscoveredReaders,
      log,
      initialize: _initialize,
      setUserCallbacks,
      emitter,
    }),
    [
      _initialize,
      loading,
      isInitialized,
      connectedReader,
      discoveredReaders,
      setIsInitialized,
      setLoading,
      setConnectedReader,
      setDiscoveredReaders,
      log,
      setUserCallbacks,
    ]
  );

  return (
    <StripeTerminalContext.Provider value={value}>
      {children}
    </StripeTerminalContext.Provider>
  );
}
