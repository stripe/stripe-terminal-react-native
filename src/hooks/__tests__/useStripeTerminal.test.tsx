import * as React from 'react';
import { useStripeTerminal } from '../useStripeTerminal';
import { act, renderHook } from '@testing-library/react-native';
import { StripeTerminalContext } from '../../components/StripeTerminalContext';
import * as functions from '../../functions';

jest.mock(
  '../../../node_modules/react-native/Libraries/EventEmitter/NativeEventEmitter'
);

function spyAllFunctions({ returnWith = null }: { returnWith?: any } = {}) {
  const createSetupIntent = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'createSetupIntent')
    .mockImplementation(createSetupIntent);
  //
  const connectReader = jest.fn(() => returnWith);
  jest.spyOn(functions, 'connectReader').mockImplementation(connectReader);
  //
  const discoverReaders = jest.fn(() => returnWith);
  jest.spyOn(functions, 'discoverReaders').mockImplementation(discoverReaders);
  //
  const cancelDiscovering = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'cancelDiscovering')
    .mockImplementation(cancelDiscovering);
  //
  const createPaymentIntent = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'createPaymentIntent')
    .mockImplementation(createPaymentIntent);
  //
  const collectPaymentMethod = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'collectPaymentMethod')
    .mockImplementation(collectPaymentMethod);
  //
  const retrievePaymentIntent = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'retrievePaymentIntent')
    .mockImplementation(retrievePaymentIntent);
  //
  const getLocations = jest.fn(() => returnWith);
  jest.spyOn(functions, 'getLocations').mockImplementation(getLocations);
  //
  const confirmPaymentIntent = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'confirmPaymentIntent')
    .mockImplementation(confirmPaymentIntent);
  //
  const cancelPaymentIntent = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'cancelPaymentIntent')
    .mockImplementation(cancelPaymentIntent);
  //
  const disconnectReader = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'disconnectReader')
    .mockImplementation(disconnectReader);
  //
  const rebootReader = jest.fn(() => returnWith);
  jest.spyOn(functions, 'rebootReader').mockImplementation(rebootReader);
  //
  const installAvailableUpdate = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'installAvailableUpdate')
    .mockImplementation(installAvailableUpdate);
  //
  const cancelInstallingUpdate = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'cancelInstallingUpdate')
    .mockImplementation(cancelInstallingUpdate);
  //
  const setReaderDisplay = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'setReaderDisplay')
    .mockImplementation(setReaderDisplay);

  //
  const clearReaderDisplay = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'clearReaderDisplay')
    .mockImplementation(clearReaderDisplay);

  //
  const retrieveSetupIntent = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'retrieveSetupIntent')
    .mockImplementation(retrieveSetupIntent);

  //
  const collectSetupIntentPaymentMethod = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'collectSetupIntentPaymentMethod')
    .mockImplementation(collectSetupIntentPaymentMethod);

  //
  const cancelSetupIntent = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'cancelSetupIntent')
    .mockImplementation(cancelSetupIntent);

  //
  const confirmSetupIntent = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'confirmSetupIntent')
    .mockImplementation(confirmSetupIntent);

  //
  const simulateReaderUpdate = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'simulateReaderUpdate')
    .mockImplementation(simulateReaderUpdate);

  //
  const collectRefundPaymentMethod = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'collectRefundPaymentMethod')
    .mockImplementation(collectRefundPaymentMethod);

  //
  const confirmRefund = jest.fn(() => returnWith);
  jest.spyOn(functions, 'confirmRefund').mockImplementation(confirmRefund);

  //
  const cancelCollectPaymentMethod = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'cancelCollectPaymentMethod')
    .mockImplementation(cancelCollectPaymentMethod);

  //
  const cancelCollectRefundPaymentMethod = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'cancelCollectRefundPaymentMethod')
    .mockImplementation(cancelCollectRefundPaymentMethod);

  //
  const cancelCollectSetupIntent = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'cancelCollectSetupIntent')
    .mockImplementation(cancelCollectSetupIntent);

  //
  const setSimulatedCard = jest.fn(() => returnWith);
  jest
    .spyOn(functions, 'setSimulatedCard')
    .mockImplementation(setSimulatedCard);

  return {
    discoverReaders,
    cancelDiscovering,
    connectReader,
    disconnectReader,
    rebootReader,
    createPaymentIntent,
    collectPaymentMethod,
    retrievePaymentIntent,
    getLocations,
    confirmPaymentIntent,
    createSetupIntent,
    cancelPaymentIntent,
    installAvailableUpdate,
    cancelInstallingUpdate,
    setReaderDisplay,
    clearReaderDisplay,
    retrieveSetupIntent,
    collectSetupIntentPaymentMethod,
    cancelSetupIntent,
    confirmSetupIntent,
    simulateReaderUpdate,
    collectRefundPaymentMethod,
    confirmRefund,
    cancelCollectPaymentMethod,
    cancelCollectRefundPaymentMethod,
    cancelCollectSetupIntent,
    setSimulatedCard,
  };
}

const createContextWrapper =
  (providerProps: any): React.FC =>
  ({ children }) =>
    (
      <StripeTerminalContext.Provider
        value={{
          isInitialized: false,
          loading: false,
          discoveredReaders: [],
          setConnectedReader: jest.fn(),
          log: jest.fn(),
          setDiscoveredReaders: jest.fn(),
          setIsInitialized: jest.fn(),
          setLoading: jest.fn(),
          connectedReader: null,
          emitter: undefined,
          initialize: jest.fn(),
          ...providerProps,
        }}
      >
        {children}
      </StripeTerminalContext.Provider>
    );

describe('useStripeTerminal.test.tsx', () => {
  describe('Public API snapshot', () => {
    it('ensure there are no unexpected changes to the hook exports', () => {
      const { result } = renderHook(() => useStripeTerminal(), {
        wrapper: createContextWrapper({}),
      });

      expect(result).toMatchSnapshot();
    });
  });

  it('should use context values', () => {
    const { result } = renderHook(() => useStripeTerminal(), {
      wrapper: createContextWrapper({
        isInitialized: true,
        loading: false,
        connectedReader: { id: 12 },
        discoveredReaders: [{ id: 12 }, { id: 15 }],
      }),
    });

    const { isInitialized, loading, connectedReader, discoveredReaders } =
      result.current;

    expect(isInitialized).toEqual(true);
    expect(loading).toEqual(loading);
    expect(connectedReader).toMatchObject({ id: 12 });
    expect(discoveredReaders).toMatchObject([{ id: 12 }, { id: 15 }]);
  });

  describe('Public methods are called properly', () => {
    it('clearCachedCredentials is called', () => {
      const clearCachedCredentials = jest.fn();
      jest
        .spyOn(functions, 'clearCachedCredentials')
        .mockImplementation(clearCachedCredentials);

      const ContextWrapper = createContextWrapper({});
      const { result } = renderHook(() => useStripeTerminal(), {
        wrapper: ContextWrapper,
      });

      act(() => {
        result.current.clearCachedCredentials();
      });

      expect(clearCachedCredentials).toBeCalled();
    });

    it('initialized method is called', () => {
      const initializeFn = jest.fn();
      const ContextWrapper = createContextWrapper({ initialize: initializeFn });
      const { result } = renderHook(() => useStripeTerminal(), {
        wrapper: ContextWrapper,
      });

      act(() => {
        result.current.initialize();
      });

      expect(initializeFn).toBeCalled();
    });

    it('public methods are called when it is initialized', () => {
      const fns = spyAllFunctions({ returnWith: '_value' });

      const ContextWrapper = createContextWrapper({ isInitialized: true });
      const { result } = renderHook(() => useStripeTerminal(), {
        wrapper: ContextWrapper,
      });

      act(() => {
        try {
          result.current.discoverReaders({} as any);
          result.current.cancelCollectPaymentMethod();
          result.current.cancelDiscovering();
          result.current.cancelCollectRefundPaymentMethod();
          result.current.cancelInstallingUpdate();
          result.current.cancelPaymentIntent({} as any);
          result.current.cancelSetupIntent({} as any);
          result.current.clearCachedCredentials();
          result.current.clearReaderDisplay();
          result.current.collectPaymentMethod({} as any);
          result.current.collectRefundPaymentMethod({} as any);
          result.current.collectSetupIntentPaymentMethod({} as any);
          result.current.confirmSetupIntent({} as any);
          result.current.connectReader({} as any, {} as any);
          result.current.createPaymentIntent({} as any);
          result.current.createSetupIntent({} as any);
          result.current.disconnectReader();
          result.current.rebootReader();
          result.current.retrievePaymentIntent('');
          result.current.getLocations({} as any);
          result.current.confirmPaymentIntent({} as any);
          result.current.retrieveSetupIntent('');
          result.current.simulateReaderUpdate({} as any);
          result.current.setSimulatedCard('');
          result.current.installAvailableUpdate();
          result.current.setReaderDisplay({} as any);
          result.current.confirmRefund();
          result.current.cancelCollectSetupIntent();
        } catch (error) {
          console.error(error);
        }
      });

      Object.values(fns).forEach((fn) => {
        expect(fn).toBeCalled();
      });
    });

    it('public methods are not called when it is not initialized', async () => {
      const fns = spyAllFunctions();
      console.error = jest.fn();

      const ContextWrapper = createContextWrapper({ isInitialized: false });
      const { result } = renderHook(() => useStripeTerminal(), {
        wrapper: ContextWrapper,
      });

      try {
        await result.current.connectReader({} as any, {} as any);
        await result.current.discoverReaders({} as any);
        await result.current.cancelCollectPaymentMethod();
        await result.current.cancelDiscovering();
        await result.current.cancelCollectRefundPaymentMethod();
        await result.current.cancelInstallingUpdate();
        await result.current.cancelPaymentIntent({} as any);
        await result.current.cancelSetupIntent({} as any);
        await result.current.clearReaderDisplay();
        await result.current.collectPaymentMethod({} as any);
        await result.current.collectRefundPaymentMethod({} as any);
        await result.current.collectSetupIntentPaymentMethod({} as any);
        await result.current.confirmSetupIntent({} as any);
        await result.current.connectReader({} as any, {} as any);
        await result.current.createPaymentIntent({} as any);
        await result.current.createSetupIntent({} as any);
        await result.current.disconnectReader();
        await result.current.rebootReader();
        await result.current.retrievePaymentIntent('');
        await result.current.getLocations({} as any);
        await result.current.confirmPaymentIntent({} as any);
        await result.current.retrieveSetupIntent('');
        await result.current.simulateReaderUpdate({} as any);
        await result.current.setSimulatedCard('');
        await result.current.installAvailableUpdate();
        await result.current.setReaderDisplay({} as any);
        await result.current.confirmRefund();
        await result.current.cancelCollectSetupIntent();
      } catch (error) {
        expect(error).toBeInstanceOf(Error);
        if (error instanceof Error) {
          expect(error.message).toEqual(
            'First initialize the Stripe Terminal SDK before performing any action'
          );
        }
      }

      Object.values(fns).forEach((fn) => {
        expect(fn).not.toBeCalled();
      });
      expect(console.error).toBeCalledWith(
        'First initialize the Stripe Terminal SDK before performing any action'
      );
      expect(console.error).toBeCalledTimes(1);
    });

    it('public methods are returns with mocked value', async () => {
      spyAllFunctions({ returnWith: '_value' });

      const ContextWrapper = createContextWrapper({ isInitialized: true });
      const { result } = renderHook(() => useStripeTerminal(), {
        wrapper: ContextWrapper,
      });

      await expect(
        result.current.connectReader({} as any, {} as any)
      ).resolves.toEqual('_value');
      await expect(result.current.discoverReaders({} as any)).resolves.toEqual(
        '_value'
      );
      await expect(
        result.current.cancelCollectPaymentMethod()
      ).resolves.toEqual('_value');
      await expect(result.current.cancelDiscovering()).resolves.toEqual(
        '_value'
      );
      await expect(
        result.current.cancelCollectRefundPaymentMethod()
      ).resolves.toEqual('_value');
      await expect(result.current.cancelInstallingUpdate()).resolves.toEqual(
        '_value'
      );
      await expect(
        result.current.cancelPaymentIntent({} as any)
      ).resolves.toEqual('_value');
      await expect(
        result.current.cancelSetupIntent({} as any)
      ).resolves.toEqual('_value');
      await expect(result.current.clearReaderDisplay()).resolves.toEqual(
        '_value'
      );
      await expect(
        result.current.collectPaymentMethod({} as any)
      ).resolves.toEqual('_value');
      await expect(
        result.current.collectRefundPaymentMethod({} as any)
      ).resolves.toEqual('_value');
      await expect(
        result.current.collectSetupIntentPaymentMethod({} as any)
      ).resolves.toEqual('_value');
      await expect(
        result.current.confirmSetupIntent({} as any)
      ).resolves.toEqual('_value');
      await expect(
        result.current.createPaymentIntent({} as any)
      ).resolves.toEqual('_value');
      await expect(
        result.current.createSetupIntent({} as any)
      ).resolves.toEqual('_value');
      await expect(result.current.disconnectReader()).resolves.toEqual(
        '_value'
      );
      await expect(result.current.rebootReader()).resolves.toEqual('_value');
      await expect(
        result.current.retrievePaymentIntent({} as any)
      ).resolves.toEqual('_value');
      await expect(result.current.getLocations({} as any)).resolves.toEqual(
        '_value'
      );
      await expect(
        result.current.confirmPaymentIntent({} as any)
      ).resolves.toEqual('_value');
      await expect(
        result.current.retrieveSetupIntent({} as any)
      ).resolves.toEqual('_value');
      await expect(
        result.current.simulateReaderUpdate({} as any)
      ).resolves.toEqual('_value');
      await expect(result.current.setSimulatedCard({} as any)).resolves.toEqual(
        '_value'
      );
      await expect(result.current.installAvailableUpdate()).resolves.toEqual(
        '_value'
      );
      await expect(result.current.setReaderDisplay({} as any)).resolves.toEqual(
        '_value'
      );
      await expect(result.current.confirmRefund()).resolves.toEqual('_value');
      await expect(result.current.cancelCollectSetupIntent()).resolves.toEqual(
        '_value'
      );
    });
  });
});
