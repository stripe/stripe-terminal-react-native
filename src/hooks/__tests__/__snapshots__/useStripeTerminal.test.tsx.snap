// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStripeTerminal.test.tsx Public API snapshot ensure there are no unexpected changes to the hook exports 1`] = `
Object {
  "current": Object {
    "cancelCollectInputs": [Function],
    "cancelCollectPaymentMethod": [Function],
    "cancelCollectRefundPaymentMethod": [Function],
    "cancelCollectSetupIntent": [Function],
    "cancelDiscovering": [Function],
    "cancelInstallingUpdate": [Function],
    "cancelPaymentIntent": [Function],
    "cancelReaderReconnection": [Function],
    "cancelSetupIntent": [Function],
    "clearCachedCredentials": [Function],
    "clearReaderDisplay": [Function],
    "collectData": [Function],
    "collectInputs": [Function],
    "collectPaymentMethod": [Function],
    "collectRefundPaymentMethod": [Function],
    "collectSetupIntentPaymentMethod": [Function],
    "confirmPaymentIntent": [Function],
    "confirmRefund": [Function],
    "confirmSetupIntent": [Function],
    "connectReader": [Function],
    "connectedReader": null,
    "createPaymentIntent": [Function],
    "createSetupIntent": [Function],
    "disconnectReader": [Function],
    "discoverReaders": [Function],
    "discoveredReaders": Array [],
    "emitter": undefined,
    "getConnectedReader": [Function],
    "getConnectionStatus": [Function],
    "getLocations": [Function],
    "getNativeSdkVersion": [Function],
    "getOfflineStatus": [Function],
    "getPaymentStatus": [Function],
    "getReaderSettings": [Function],
    "initialize": [Function],
    "installAvailableUpdate": [Function],
    "isInitialized": false,
    "loading": false,
    "rebootReader": [Function],
    "retrievePaymentIntent": [Function],
    "retrieveSetupIntent": [Function],
    "setReaderDisplay": [Function],
    "setReaderSettings": [Function],
    "setSimulatedCard": [Function],
    "setTapToPayUxConfiguration": [Function],
    "simulateReaderUpdate": [Function],
    "supportsReadersOfType": [Function],
  },
}
`;
